compiler/src/iree/compiler/Codegen/Common/MaterializeEncodingIntoNop.cpp

```diff
@@ -24,6 +24,25 @@ namespace mlir::iree_compiler {
 using namespace IREE::Encoding;
 
 namespace {
+
+struct FoldAwayExtractSliceOnUnsetEncodingPattern
+    : public OpConversionPattern<tensor::ExtractSliceOp> {
+  using OpConversionPattern<tensor::ExtractSliceOp>::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(tensor::ExtractSliceOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    auto unsetEncodingOp =
+        op.getSource().getDefiningOp<IREE::Encoding::UnsetEncodingOp>();
+    if (!unsetEncodingOp) {
+      return rewriter.notifyMatchFailure(op,
+                                         "source is not an unset_encoding op");
+    }
+    rewriter.replaceOp(op, adaptor.getSource());
+    return success();
```

Comment:
Is it possible for the `extract_slice` which is the consumer of the `unset_encoding` to have changed? For example:

```mlir
%4 = iree_encoding.unset_encoding %3 : tensor<?x16384xf32, #encoding_acc> -> tensor<?x16384xf32>
%extracted_slice = tensor.extract_slice %4[0, 0] [%dim, 16384] [1, 1] : tensor<?x16384xf32> to tensor<?x16384xf32>
%extracted_slice_2 = tensor.extract_slice %extracted_slice[0, 0] [%dim_2, 16384] [1, 1] : tensor<?x16384xf32> to tensor<?x16384xf32>
```

The first `tensor.extract_slice` is redundant because `%dim <= %dim_2` and all the static dims are equal. So it could get folded to:

```mlir
%4 = iree_encoding.unset_encoding %3 : tensor<?x16384xf32, #encoding_acc> -> tensor<?x16384xf32>
%extracted_slice_2 = tensor.extract_slice %extracted_slice[0, 0] [%dim_2, 16384] [1, 1] : tensor<?x16384xf32> to tensor<?x16384xf32>
```

And if `%dim_2` is strictly less than `%dim` it would be incorrect to erase the extract slice.

Reply:
The solution is built on top of how we set up encodings. Given that the pass aims to drop the encodings (i.e., turn them into nop), I think we can assume that it knows how it is created. Thus, the example that you provided should be from below, which means that cancelling the extract_slice should still be fine. I agree that it is not that generic, but the solution looks fine to me. Because they are really data-tiling specific at this moment.

```mlir
%0 = linalg.matmul ...
%1 = extract_slice %0 ...
```

Reply:
oh, there are other issues might pop out when they are static shapes..

Reply:
I made a change and add a comment describe that the pattern is fragile in some setup, but I think it should be okay once we move to the recent data-tiling path that we're building.